Sources/SwiftProtobuf/SwiftProtobufError.swift

```diff
@@ -0,0 +1,1098 @@
+
+/// A SwiftProtobuf specific error.
+///
+/// All errors have a high-level ``SwiftProtobufError/Code-swift.struct`` which identifies the domain
+/// of the error. For example, an issue when encoding a proto into binary data will result in a
+/// ``SwiftProtobufError/Code-swift.struct/binaryEncodingError`` error code.
+/// Errors also include a message describing what went wrong and how to remedy it (if applicable). The
+/// ``SwiftProtobufError/message`` is not static and may include dynamic information such as the
+/// type URL for a type that could not be decoded, for example.
+public struct SwiftProtobufError: Error, @unchecked Sendable {
+    // Note: @unchecked because we use a backing class for storage.
+
+    private var storage: Storage
+    private mutating func ensureStorageIsUnique() {
+        if !isKnownUniquelyReferenced(&self.storage) {
+            self.storage = self.storage.copy()
+        }
+    }
+    
+    private final class Storage {
+        var code: Code
+        var message: String
+        var location: SourceLocation
+
+        init(
+            code: Code,
+            message: String,
+            location: SourceLocation
+        ) {
+            self.code = code
+            self.message = message
+            self.location = location
+        }
+
+        func copy() -> Self {
+            return Self(
+                code: self.code,
+                message: self.message,
+                location: self.location
+            )
+        }
+    }
+    
+    /// A high-level error code to provide broad a classification.
+    public var code: Code {
+        get { self.storage.code }
+        set {
+            self.ensureStorageIsUnique()
+            self.storage.code = newValue
+        }
+    }
+
+    /// A message describing what went wrong and how it may be remedied.
+    internal var message: String {
+        get { self.storage.message }
+        set {
+            self.ensureStorageIsUnique()
+            self.storage.message = newValue
+        }
+    }
+    
+    private var location: SourceLocation {
+        get { self.storage.location }
+        set {
+            self.ensureStorageIsUnique()
+            self.storage.location = newValue
+        }
+    }
+
+    public init(
+        code: Code,
+        message: String,
+        location: SourceLocation
+    ) {
+        self.storage = Storage(code: code, message: message, location: location)
+    }
+}
+
+extension SwiftProtobufError {
+    /// A high level indication of the kind of error being thrown.
+    public struct Code: Hashable, Sendable, CustomStringConvertible {
+        private enum Wrapped: Hashable, Sendable, CustomStringConvertible {
+            case binaryEncodingError
+            case binaryDecodingError
+            case binaryStreamDecodingError
+            case jsonEncodingError
+            case jsonDecodingError
+            case textFormatDecodingError
+            case invalidArgument
+            case internalError
```

Comment:
So what's the recommendation going forward for when to add more codes?  If folks now can use the codes to figure out error types, should we keep the previous granularity?

Reply:
What we've done in other open source projects is to have this code represent a domain, instead of a particular error - the rationale is that callers should be able to handle all errors from the same domain equally, without having so much unnecessary granularity that it ends up being a burden for users. 

For example, now all errors thrown when decoding JSON relating to the format of the received payload will have a `jsonDecodingError` code; errors relating to problems with some internal representations or other internal unrecoverable state will throw `internalError`; errors relating to failed binary encoding because of issues with the given message will throw `binaryEncodingError`, etc...

I followed this pattern here as well, which means you'd only add new cases here if a new domain category is warranted - other you'd reuse one of the existing codes for new errors if applicable. 

Does this make sense or have you got any concerns/questions?

Reply:
Make sense.